src/Client.php

```diff
@@ -105,33 +131,64 @@ public function runQuery($query, bool $resultsAsArray = false, array $variables
      */
     public function runRawQuery(string $queryString, $resultsAsArray = false, array $variables = []): Results
     {
-        $request = new Request($this->requestMethod, $this->endpointUrl);
-
-        foreach($this->httpHeaders as $header => $value) {
-            $request = $request->withHeader($header, $value);
-        }
+        $request = $this->requestFactory->createRequest($this->requestMethod, $this->endpointUrl);
 
         // Convert empty variables array to empty json object
         if (empty($variables)) $variables = (object) null;
+
         // Set query in the request body
         $bodyArray = ['query' => (string) $queryString, 'variables' => $variables];
-        $request = $request->withBody(Psr7\stream_for(json_encode($bodyArray)));
 
-        // Send api request and get response
+        $stream = $this->streamFactory->createStream(json_encode($bodyArray));
+
+        $request = (new RequestDecorator())->decorate($request, $stream, $this->httpOptions, $this->httpHeaders);
+
         try {
+            $request = $request->withBody($stream);
+
+            foreach($this->httpHeaders as $header => $value) {
+                $request = $request->withHeader($header, $value);
+            }
+
+            // Send api request and get response
             $response = $this->httpClient->sendRequest($request);
         }
-        catch (ClientException $exception) {
-            $response = $exception->getResponse();
+        catch (BaseInvalidArgumentException $e) {
```

Comment:
I can't find anything in the PSR-18 which states that a compliant HTTP client can throw an `InvalidArgumentException`, why are we catching this?

Reply:
I know that Guzzle implements PSR-18 and will throw it. But thinking about it, I can see how it would couple the library with Guzzle or something that throws it. We can remove this.